### PR TITLE
Allow Symfony 6

### PR DESCRIPTION
**Pull request by Kocal**

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
We should wait a bit before allowing Symfony 6 because we're still breaking BC there (eg adding return types).

**Comment by greg0ire** [comment]
Converting to draft until the right moment.

**Comment by wouterj** [comment]
Fyi, all types have been added and work on this PR can resume :)

**Review by derrabus — APPROVED** [review]

**Comment by derrabus** [comment]
Can you please rebase and resolve the conflicts?

**Comment by Kocal** [comment]
Done, however I think it won't works because I can't set Composer's `minimum-stability` anymore.
 
Should we wait for Symfony 6 official release to re-run this PR's checks?

**Comment by derrabus** [comment]
> Done, however I think it won't works because I can't set Composer's `minimum-stability` anymore.

@greg0ire Can you have a look, please?

**Comment by greg0ire** [comment]
I've tried something, but it's getting late, and I have to think this through a bit more (the extra prefer-lowest job and upload coverage job seem wrong)

**Comment by greg0ire** [comment]
I removed my commit because I cannot find a satisfying solution. I think in this case, we should partially revert 739d439, or just maintain an MR open that sets the stability in composer.json to dev, and merge it (after dropping the commit that does this) once Symfony 6 comes out. Having a workflow that oscillates between a custom version and a generic one does not sound great.

**Comment by derrabus** [comment]
@greg0ire But adjusting the minimum stability for a CI job is not a very uncommon need. Wouldn't it be possible to add an optional parameter to the workflow that would result in a `composer config minimum-stability $value` being run before installing the dependencies? That's all we need here, I guess.

File: composer.json [hunk 1]
```diff
@@ -21,10 +21,11 @@
     ],
     "require": {
         "php": "^7.2|^8.0",
-        "symfony/framework-bundle": "~3.4|~4.0|~5.0",
+        "symfony/framework-bundle": "~3.4|~4.0|~5.0|6.0.x@dev",
```
Comment (derrabus) [inline composer.json]:
```suggestion
        "symfony/framework-bundle": "~3.4|~4.0|~5.0|~6.0",
```

This is redundant if you lower the minimum stability globally. 😉

**Comment by greg0ire** [comment]
> But adjusting the minimum stability for a CI job is not a very uncommon need. Wouldn't it be possible to add an optional parameter to the workflow that would result in a composer config minimum-stability $value being run before installing the dependencies? That's all we need here, I guess.

We would also need an extra job under `include` in the build matrix, and currently there does not seem to be a straightforward way to make it conditional (I'm assuming we don't want to do this for every Doctrine repository, can be challenged at the org level).

File: composer.json [hunk 2]
```diff
@@ -21,10 +21,11 @@
     ],
     "require": {
         "php": "^7.2|^8.0",
-        "symfony/framework-bundle": "~3.4|~4.0|~5.0",
+        "symfony/framework-bundle": "~3.4|~4.0|~5.0|~6.0",
         "doctrine/doctrine-bundle": "~1.0|~2.0",
         "doctrine/migrations": "^3.2"
     },
+    "minimum-stability": "dev",
```
Comment (derrabus) [inline composer.json]: We could try `beta` here, if we don't want a bleeding edge Psalm in our pipline. 🤔 

```suggestion
    "minimum-stability": "beta",
```

**Comment by derrabus** [comment]
> We would also need an extra job under `include` in the build matrix, and currently there does not seem to be a straightforward way to make it conditional

I see.

Would it be possible to decouple the workflow from the matrix? My idea would be to have one parameterized `phpunit` workflow that runs the tests for a single given `php-version` and `dependencies` setting. The `continuous-integration` workflow run this new `phpunit` workflow with different parameters.

This way, we could add additional parameters to the `phpunit` workflow that the `continuous-integration` matrix does not make use of, like `minimum-stablity`. A project could use the `continuous-integration` workflow and add additional `phpunit` jobs with custom settings.

File: composer.json (hunk 2 above)
Comment (kbond) [reply composer.json]: I believe adding `"prefer-stable": true` would help.

Comment (derrabus) [reply composer.json]: @kbond No, because that would install Symfony 5 packages on the PHP 8.0 run which is exactly not what we want.

**Comment by derrabus** [comment]
Psalm's `ReservedWord` check is broken. It complains because Symfony has added `mixed` type declarations to parameters. I'll prepare a bug report for Psalm, but for now I've disabled it because it reports false positives.

**Comment by greg0ire** [comment]
> Would it be possible to decouple the workflow from the matrix? My idea would be to have one parameterized phpunit workflow that runs the tests for a single given php-version and dependencies setting. The continuous-integration workflow run this new phpunit workflow with different parameters.

> This way, we could add additional parameters to the phpunit workflow that the continuous-integration matrix does not make use of, like minimum-stablity. A project could use the continuous-integration workflow and add additional phpunit jobs with custom settings.

Yes, that would be possible and was in fact quite similar to what I did with my previous attempt. There are things left to get right though, like how to upload coverage. I'll give it another try.

**Comment by greg0ire** [comment]
My new attempt can be seen at https://github.com/doctrine/DoctrineMigrationsBundle/pull/456

**Pull request merged**